### ISSUE TITLE
bus error on aarch64 mmap to PCIE bar memory

### ISSUE THREAD

**Issue by celeryfake** [jemalloc/jemalloc#801]
Hi all,
Currently I was using jemalloc on a arm64 testbed. I am trying to using jemalloc to allocate memory from PCIE bar memory through malloc function, basically make a memory pool. Currently I am using version 4.4.0.
What I did is changed the pages_map() function in pages.c to mmap from PCIE bar memory instead of host DRAM(anonymous pages). I wrote a driver to let, the PCIE bar memory accessed through cache, so there is no alignment issue rd/wr on memory mmap area on PCIE bar memory. In order to make the pointer return by mmap 2MB aligned, I used MAP_FIXED with aligned virtual address, so it never need to re-mmaped. The physical memory mmap on PCIE bar is managed in the driver side.

But when I tested it on the test bed. I got this weird bus error, here is the call stack:
Program received signal SIGBUS, Bus error.
je_atomic_add_u (p=0x7bd, x=x@entry=0)
    at include/jemalloc/internal/atomic.h:608   
608     include/jemalloc/internal/atomic.h: No such file or directory.
(gdb) backtrace
#0  je_atomic_add_u (p=0x7bd, x=x@entry=0)
    at include/jemalloc/internal/atomic.h:608   
#1  0x0000ffffb7f9be2c in je_arena_nthreads_get (arena=<optimized out>,
    internal=internal@entry=false) at src/arena.c:3516
#2  0x0000ffffb7f915ec in je_arena_choose_hard (tsd=0xffffb7ff7700,
    internal=255) at src/jemalloc.c:636
#3  0x0000ffffb7f91848 in je_arena_choose_impl (tsd=<optimized out>,
    internal=<optimized out>, arena=0x0)
    at include/jemalloc/internal/jemalloc_internal.h:903
#4  0x0000ffffb7f91870 in je_arena_choose_impl (internal=false,
    arena=arena@entry=0x0, tsd=tsd@entry=0xffffb7ff7700)
    at include/jemalloc/internal/jemalloc_internal.h:913
#5  je_arena_choose (tsd=tsd@entry=0xffffb7ff7700, arena=arena@entry=0x0)
    at include/jemalloc/internal/jemalloc_internal.h:912
#6  0x0000ffffb7fb3d90 in je_tcache_get_hard (tsd=0xffffb7ff7700)
    at src/tcache.c:311
#7  0x0000ffffb7fb3e1c in je_tcache_get (tsd=tsd@entry=0xffffb7ff7700,
    create=create@entry=true) at include/jemalloc/internal/tcache.h:234
#8  0x0000ffffb7f8f358 in je_ialloc (tsd=0xffffb7ff7700, size=143, ind=9,
    zero=false, slow_path=true)
    at include/jemalloc/internal/jemalloc_internal.h:1073
#9  0x0000ffffb7f90b0c in ialloc_body (size=<optimized out>,
    zero=<optimized out>, tsdn=<optimized out>, usize=<optimized out>,
---Type <return> to continue, or q <return> to quit---
    slow_path=<optimized out>) at src/jemalloc.c:1612
#10 0x0000ffffb7f90b80 in malloc (size=<optimized out>) at src/jemalloc.c:1654
#11 0x000000000040090c in main (argc=1, argv=0xfffffffffc38)
    at test_malloc.c:27

I tried to make sure my modification on the mmap doesn't cause the problem, I changed the mmap to PCIE part to mmap to anonymous pages on host DRAM (still using MAP_FIXED). It seems works fine on the host DRAM. I suspect that the library is working on host DRAM but the mmap memory is working on the PCIE bar memory, this may cause the issue.

Can anybody give me some help, I am not very familiar with the detail implementation of jemalloc. I am working on that right now.

Thanks and best regards

**Comment by interwq** [comment]
Something looks messed up. In this frame:
#2 0x0000ffffb7f915ec in je_arena_choose_hard (tsd=0xffffb7ff7700,
internal=255) at src/jemalloc.c:636

Internal should be a boolean and this line number doesn't look like 4.4.0, https://github.com/jemalloc/jemalloc/blob/4.4.0/src/jemalloc.c#L636

Can you check if you are using the right version?

**Comment by celeryfake** [comment]
Thanks for the comment!

I did use the 4.4.0 version. The line number wrong is because I changed the jemalloc.c and pages.c to hook up with my driver to mmap memory from PCIE bar memory.

And the internal parameter in #2 0x0000ffffb7f915ec in je_arena_choose_hard (tsd=0xffffb7ff7700,
internal=255) at src/jemalloc.c:636 did seem wrong in the program logic.

I have a version works fine with mmap arm host DRAM (mmap anonymous page). I compared the program traces of the two version. The difference of program order starts from here: 

The arena_get in the if statement get an non-zero pointer which leads to the bus error (je_atomic_add_u (p=0x7bd, x=x@entry=0))

arena_choose_hard at src/jemalloc.c:636
if (arena_get(tsd_tsdn(tsd), i, false) != NULL) {
				/*
				 * Choose the first arena that has the lowest
				 * number of threads assigned to it.
				 */
				for (j = 0; j < 2; j++) {
					if (arena_nthreads_get(arena_get(
					    tsd_tsdn(tsd), i, false), !!j) <
					    arena_nthreads_get(arena_get(
					    tsd_tsdn(tsd), choose[j], false),
					    !!j))
						choose[j] = i;

I was curious about how the global variable arena_t **arenas is used. Because I think the error comes from that variable (in the arena_get function it return a non-zero value which is the illegal address it access that lead to the bus error).

Thanks and best regard

**Comment by interwq** [comment]
@celeryfake : sounds like this is memory corruption caused by the changes. ```arena_t **arenas ``` points to an array which contains pointers to all available arenas. To allocate from the customized memory range, have you tried allocation hooks? http://jemalloc.net/jemalloc.3.html#arena.i.chunk_hooks

I'd suggest using manual arenas (http://jemalloc.net/jemalloc.3.html#arenas.extend) and chunk hooks above to manage the PCIE memory. From what you described I believe it should have support you need.

**Comment by celeryfake** [comment]
Actually I have tried that in the first place, but only on x86 platfrom. It works fine. But the problem of that is we need to change the application code. That's why we dropped this idea and tried to modified the jemalloc source code to hook up with the device driver. Any other way to get it around?

Thanks

**Comment by celeryfake** [comment]
Now I found the problem. 

The issue is caused by base_alloc() function which initialize the **arenas. It suppose "demand-zeroed memory".  While for mmap anonymous page which is used by default mmap, the kernel guarantees that using something like COW. However, when I changed it to mmap the PCIE bar memory using my driver, it read data as it is on the device memory. That cause the issue, different initial data on the PCIE device memory may cause different error actually.

When I write PCIE bar memory all zero before calling the jemalloc library, it seems working fine. 

But I still need some more testing to see whether there is other issues.

Thanks for the help
